### DISCUSSION TITLE
Make local testing of the code base more reliable and reproducible (tox?)

### DISCUSSION THREAD

**Discussion by ssbarnea** [mkdocstrings/mkdocstrings#516] [Ideas]
Based on my experience trying to contribute some minor bugfixes to mkdocstrings, I faces a set of barriers which should be considered as they are likely to deter others from contributing.

Despite the fact that the project is a pure-python project, is not using `tox` for its testing. Instead it is using a combination of `pdm` and `make`. I installed pdb just to find out that the tool was crashing on my system due to a 2 months old bug https://github.com/pdm-project/pdm/issues/1510 - this required few more hours to engage with these projects to fix that part.

After that, I realised that even running `make check` (supposedly linting) was failing even on the clean cloned code:
```
ssbarnea@m1: ~/c/os/mkdocs/mkdocstrings master
$ make check
> Currently selected Python version: python
> pdm run duty check-quality check-types check-docs (python3.7)
✗ Checking code quality (1)
  > flake8 --config=config/flake8.ini src tests duties.py docs
  src/mkdocstrings/loggers.py:1:1: FA100 Missing from __future__ import annotations but imports: Optional, Tuple
  src/mkdocstrings/extension.py:1:1: FA100 Missing from __future__ import annotations but imports: Tuple
  src/mkdocstrings/plugin.py:1:1: FA100 Missing from __future__ import annotations but imports: List, Optional, Tuple
  src/mkdocstrings/inventory.py:1:1: FA100 Missing from __future__ import annotations but imports: List, Optional
  src/mkdocstrings/handlers/rendering.py:1:1: FA100 Missing from __future__ import annotations but imports: Dict, List, Optional
  duties.py:1:1: FA100 Missing from __future__ import annotations but imports: List, Optional

> Restored previous Python version: python
make: *** [check] Error 1
FAIL: 2
```

I happen to know that this is caused by the fact that I i have several plugin installed in the local copy of flake8.
```
$ flake8 --version
3.9.2 (flake8-darglint: 1.8.1, flake8-docstrings: 1.6.0, pydocstyle: 6.1.1, flake8-future-annotations: 1.0.0, mccabe: 0.6.1, pycodestyle: 2.7.0, pyflakes: 2.3.1)
CPython 3.11.0+ on Darwin
```

My expectation what that the use of make/pdm was to ensure a reliable local testing experience, but apparently they don't make any use of virtualenvs for runnings tests.

So, I would likely want to suggest consideration of using tools that produce reproducible results like [pre-commit](https://pre-commit.com/) and [tox](https://tox.wiki/en/latest/config.html) as uploading changes to GHA to test code or even lint, makes contribution harder.

Basically in less than a day I found to notable technical barriers in contributing and I do not know if these are the only ones. At least the human interaction was very pleasant so far.

I know that in the end, each maintainer has the freedom to choose his own poison, but I raised this in order raise awareness. I am curious if I am alone with my view or others faces similar challenges trying to hack the code.

**Comment by ssbarnea** [comment]
Apparently pdb problems are not done, initial fix was incomplete, still chokes with 3.11-dev

**Comment by pawamoy** [comment]
I'm sorry that you had a bad experience contributing to mkdocstrings :confused: I definitely don't want that for contributors. And I thank you for expressing what you experienced in a calm and constructive manner. This is really appreciated :slightly_smiling_face: 

PDM does have issues, but they're most of the time fixed very quickly by the maintainer, @frostming. I'll throw :rose: flowers at him :rose: (hey @frostming :wink:), and say that I've never seen someone maintain a project at such a fast pace. 871 issues and 582 PRs (most from him) closed in less than 3 years, that's more than one issue/PR per day!
Concerning your particular issue, I'd go ahead and guess too few users voiced concern about it, preventing @frostming from prioritizing it accordingly.

It is true as well that CI sometimes break for mkdocstrings (and my other projects). Tools get updated and I don't set upper bounds so sometimes there are new code quality (flake8, mypy) warnings, etc. I prefer that than spending much more time setting barriers and lifting them one after the other. I do try to fix any CI issue as quick as possible though :slightly_smiling_face: Sometimes I just don't have enough time to answer the same day or even the same week.

About your flake8 plugins leaking into PDM runs, that's indeed problematic. I hope PDM could fix it (if that's a confirmed bug), or that you could isolate your flake8 installation, with pipx maybe:

```
pipx install flake8
pipx inject flake8 plugin1 plugin2 etc
```

I don't use virtualenvs myself indeed (except for quick one-off tests, or indirectly through other tools like pipx), but I think you can setup any PDM projects with virtualenvs instead of PEP 582's `__pypackages__`. From the top of my head:

```
pdm config python.use_venv true
pdm config venv.in_project false  # required to test on multiple Python versions I think

# then you can pdm install naturally and enjoy venvs
pdm install
```

About pre-commit and tox, I've used them in the past but... I don't like them :sweat_smile: Well, I did not try tox again in a long time so maybe I would like it again. But both tools as I know them really don't fit my [workflow](https://pawamoy.github.io/posts/somewhat-modern-python-development/). One of my unpopular opinions I guess :stuck_out_tongue:. I can expand if needed.

---

I'll test my workflow locally to see if there's anything to fix, and I'll keep you updated :slightly_smiling_face:

**Reply by pawamoy** [reply]
Small update: the way I was calling external tools from my tasks was actually leaking the user-installed package (like flake8 plugins). My projects now call these tools' CLI directly (in the same Python process) so that leak is eliminated. I could probably also have used Python's isolation options.

**Comment by pawamoy** [comment]
Just ran `make lock setup format check test`, no issues :confused: 
So you really had bad luck with that Python 3.11 version and flake8 installation of yours :sweat_smile:

**Comment by ssbarnea** [comment]
Related issue: use of non standard flake8 config prevents tools using flake8 from reading it, for example flake8 in vscode will complain about line length as the default value is different than the in-repo custom value. This can easily be address by avoiding the non standard config file location... basically the `flake8 --config=config/flake8.ini` is the problem.

**Reply by pawamoy** [reply]
The IDE can be configured per project to use a specific config file. I understand this is annoying for contributors. I believe I could add a script that configures popular IDEs depending on what is found in the current folder (`.idea/`, `.vscode/`, etc.), or at least document it in the contributing guide :+1:

**Comment by ssbarnea** [comment]
BTW, I think that it would be a good idea to convert this issue into a discussion. I missed to observe that the repo already had discussions enabled. Its is more of a open discussion than anything directly fixable.
